dspy/clients/lm.py

```diff
@@ -92,51 +103,86 @@ def inspect_history(self, n: int = 1):
 
     def launch(self):
-        """Send a request to the provider to launch the model, if needed."""
-        msg = f"`launch()` is called for the auto-launched model {self.model}"
-        msg += " -- no action is taken!"
-        logger.info(msg)
+        self.provider.launch(self.model, self.launch_kwargs)
 
     def kill(self):
-        """Send a request to the provider to kill the model, if needed."""
-        msg = f"`kill()` is called for the auto-launched model {self.model}"
-        msg += " -- no action is taken!"
-        logger.info(msg)
+        self.provider.kill(self.model, self.launch_kwargs)
 
     def finetune(
             self,
             train_data: List[Dict[str, Any]],
             train_kwargs: Optional[Dict[str, Any]]=None,
-            train_method: TrainingMethod = TrainingMethod.SFT,
-            provider: str = "openai",
-            cache_finetune: bool = True,
-        ) -> FinetuneJob:
-        """Start model fine-tuning, if supported."""
+            data_format: Optional[DataFormat] = None,
+        ) -> TrainingJob:
         from dspy import settings as settings
         err = "Fine-tuning is an experimental feature."
         err += " Set `dspy.settings.experimental` to `True` to use it."
         assert settings.experimental, err
 
-        FinetuneJobClass = get_provider_finetune_job_class(provider=provider)
-        finetune_job = FinetuneJobClass(
+        err = f"Provider {self.provider} does not support fine-tuning."
+        assert self.provider.finetunable, err
+
+        # Perform data validation before starting the thread to fail early
+        train_kwargs = train_kwargs or {}
+        if not data_format:
+            adapter = self.infer_adapter()
+            data_format = infer_data_format(adapter)
+        validate_data_format(data=train_data, data_format=data_format)
+
+        # TODO(PR): We can quickly add caching, but doing so requires
+        # adding functions that just call other functions as we had in the last
+        # iteration, unless people have other ideas.
+        def thread_function_wrapper():
+            return self._run_finetune_job(job)
+
+        thread = threading.Thread(target=thread_function_wrapper)
+        job = self.provider.TrainingJob(
+            thread=thread,
             model=self.model,
             train_data=train_data,
             train_kwargs=train_kwargs,
-            train_method=train_method,
-            provider=provider
+            data_format=data_format
         )
-
-        executor = ThreadPoolExecutor(max_workers=1)
-        executor.submit(
-            execute_finetune_job,
-            finetune_job,
-            lm=self,
-            cache_finetune=cache_finetune
-        )
-        executor.shutdown(wait=False)
-
-        return finetune_job
+        thread.start()
+
+        return job
+
+    def _run_finetune_job(self, job: TrainingJob):
+        # TODO(enhance): We should listen for keyboard interrupts somewhere.
```

Comment:
That's a good point -- finetune returns immediately after starting a thread, which runs this blocking function.

It would be nicer to switch all of this to by async, perhaps we should discuss for the next iteration.